### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by mp-t**
I've limited this PR to trivial fixes.
Some mem leaks, undefined behavior, copy-pasta and const refs.

### CODE REVIEW

File: rpcs3/Emu/Cell/PPUInterpreter.cpp [hunk 1]
```diff
@@ -2829,8 +2829,8 @@ bool ppu_interpreter::BC(ppu_thread& ppu, ppu_opcode_t op)
 
 	ppu.ctr -= (bo2 ^ true);
 
-	const bool ctr_ok = bo2 | ((ppu.ctr != 0) ^ bo3);
-	const bool cond_ok = bo0 | (ppu.cr[op.bi] ^ (bo1 ^ true));
+	const bool ctr_ok = bo2 || ((ppu.ctr != 0) ^ bo3);
+	const bool cond_ok = bo0 || (ppu.cr[op.bi] ^ (bo1 ^ true));
```
Comment (Farseer2) [inline rpcs3/Emu/Cell/PPUInterpreter.cpp]: These were meant to be a bitwise or

File: rpcs3/Emu/Cell/PPUInterpreter.cpp [hunk 2]
```diff
@@ -2880,8 +2880,8 @@ bool ppu_interpreter::BCLR(ppu_thread& ppu, ppu_opcode_t op)
...
```
Comment (Farseer2) [inline rpcs3/Emu/Cell/PPUInterpreter.cpp]: Same here

File: rpcs3/Emu/Cell/PPUInterpreter.cpp [hunk 3]
```diff
@@ -2904,7 +2904,7 @@ bool ppu_interpreter::CRNOR(ppu_thread& ppu, ppu_opcode_t op)
 
 bool ppu_interpreter::CRANDC(ppu_thread& ppu, ppu_opcode_t op)
 {
-	ppu.cr[op.crbd] = ppu.cr[op.crba] & (ppu.cr[op.crbb] ^ true);
+	ppu.cr[op.crbd] = ppu.cr[op.crba] && (ppu.cr[op.crbb] ^ true);
```
Comment (Farseer2) [inline rpcs3/Emu/Cell/PPUInterpreter.cpp]: And here

File: rpcs3/Emu/Cell/PPUInterpreter.cpp [hunk 4]
```diff
@@ -2940,7 +2940,7 @@ bool ppu_interpreter::CREQV(ppu_thread& ppu, ppu_opcode_t op)
 
 bool ppu_interpreter::CRORC(ppu_thread& ppu, ppu_opcode_t op)
 {
-	ppu.cr[op.crbd] = ppu.cr[op.crba] | (ppu.cr[op.crbb] ^ true);
+	ppu.cr[op.crbd] = ppu.cr[op.crba] || (ppu.cr[op.crbb] ^ true);
```
Comment (Farseer2) [inline rpcs3/Emu/Cell/PPUInterpreter.cpp]: And here

File: rpcs3/Emu/Cell/PPUInterpreter.cpp (hunk 1 above)
Comment (yiweigang) [inline rpcs3/Emu/Cell/PPUInterpreter.cpp]: The end result is the same

Comment (mp-t) [inline rpcs3/Emu/Cell/PPUInterpreter.cpp]: All involved variables are of type bool.
Same reasoning on the other occurences.

Comment (Nekotekina) [inline rpcs3/Emu/Cell/PPUInterpreter.cpp]: MSVC generates really shitty code if you use || or && operators there and all other places, although it shouldn't affect it in theory.

Comment (mp-t) [inline rpcs3/Emu/Cell/PPUInterpreter.cpp]: Nothing too wild in the assembly. The usual short-circuiting vs. added conditionals trade off. Probably not worth the time to micro-benchmark this, so I'll revert the change.

**Comment by Nekotekina** [comment]
Why you say memory leak when it's just UB in crypto?
Also I believe "++x += y;" isn't UB, when I'm not against the change in code.

**Comment by mp-t** [comment]
`++x += y;` is indeed undefined behavior. x gets modified twice between what the old standard called sequence points. 

`std::unique_ptr<u8> metadata(new u8[metadata_size]);` is a memory leak. The referenced object gets freed with `delete` instead of `delete[]`. You need the array specific template for that.
On a more general note, `make_shared` and `make_unique` should be preferred over new.

**Comment by mp-t** [comment]
No. Only the first element of the array will be freed, the others will not.

**Comment by kd-11** [comment]
@mp-t u8 is a base data type with no destructor. Since free does not take a size variable, this has no effect on most implementations afaik. I'm not defending the mismatched scalar declaration + vector allocation, but it should not leak. Valgrind will complain about it, but in reality delete should still free the whole range (at least on gcc)

**Comment by paulsapps** [comment]
Its pointless to say its not leak because its UB, anything can happen. Otherwise you can use this argument for all other types of UB.

**Comment by kd-11** [comment]
In this case the underlying mechanism is well known (malloc/free), and stdc++ sources can confirm the same. It is UB no doubt, but I think the original complaint was that it isnt a memory leak. Its a minor issue with wording and nothing more. It cannot be proven to be a leak, but it can be proven to not be a leak. It is however undoubtedly UB, so why not word it as such.

**Pull request merged**